### PR TITLE
Add SVFA base confluence algorithm

### PR DESCRIPTION
**Pull request by rafaelmotaalves**
This implementation used the SVFAAnalysis class to execute the svfa analysis twice: detecting flows from source to base, and sink to base. This is done by overriding some methods of the SVFA to redefine the source and sink. The base lines are defined by: those that are neither source or sink, and also are on a method body

### CODE REVIEW

File: src/main/java/br/unb/cic/analysis/AbstractMergeConflictDefinition.java [hunk 1]
```diff
@@ -17,18 +19,49 @@
 public abstract class AbstractMergeConflictDefinition {
     protected List<Statement> sourceStatements;
     protected List<Statement> sinkStatements;
+    protected List<Statement> inBetweenStatements;
 
     public AbstractMergeConflictDefinition() {
         sourceStatements = new ArrayList<>();
         sinkStatements = new ArrayList<>();
+        inBetweenStatements = new ArrayList<>();
```
Comment (pauloborba) [inline src/main/java/br/unb/cic/analysis/AbstractMergeConflictDefinition.java:27]: @rafaelmotaalves what do you mean by inBetweenStatements? and why you need this? is this the same as baseStatements? do you consider that we can have left; base; right; base; left and so on?

File: src/main/java/br/unb/cic/analysis/svfa/SVFAAnalysis.java [hunk 2]
```diff
@@ -32,22 +34,23 @@ public String sootClassPath() {
     }
 
     @Override
-    public List<String> getIncludeList() {
+    public scala.collection.immutable.List<String> getIncludeList() {
```
Comment (pauloborba) [inline src/main/java/br/unb/cic/analysis/svfa/SVFAAnalysis.java:46]: @rafaelmotaalves it's not clear what's the intention of the commit. whatever it is, this doesn't seem related to the intention. so better split the commit into two next time.

File: src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import br.unb.cic.analysis.AbstractMergeConflictDefinition;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class SVFAConfluenceAnalysisWithMultipleConfluences {
```
Comment (pauloborba) [inline src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java:11]: @rafaelmotaalves is this a test? shouldn't it be in another place?

File: src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        return result;
+    }
+
+    /**
+     * @param path A list of nodes with at least 2 nodes
+     * @return The last node of the list
+     */
+    private Node getLastNode(List<Node> path) {
+        assert path.size() > 1; // assume that all paths have at least one source and one sink
```
Comment (pauloborba) [inline src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:78]: @rafaelmotaalves good! could maybe extract path.size() into a variable

File: src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+            @Override
+            protected List<Statement> getSourceStatements() {
+                return definition.getSourceStatements();
+            }
+
+            @Override
+            protected boolean isSink(Unit unit) {
+                return !isSource(unit) &&
+                        unit.getJavaSourceStartLineNumber() > 0 && // this is for ignoring lines that are not in any methods body
```
Comment (pauloborba) [inline src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:96]: @rafaelmotaalves shouldn't we assume here that the soot analysis are only called with lines that correspond to method bodies?

File: src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        return new SVFAAnalysis(this.cp, this.definition) {
+
+            @Override
+            protected List<Statement> getSourceStatements() {
+                return definition.getSourceStatements();
+            }
+
+            @Override
+            protected boolean isSink(Unit unit) {
+                return !isSource(unit) &&
```
Comment (pauloborba) [inline src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:95]: @rafaelmotaalves in principle, a unit could be marked as both source and sink. this would break the logic here

File: src/main/java/br/unb/cic/analysis/AbstractMergeConflictDefinition.java (hunk 1 above)
Comment (rafaelmotaalves) [reply src/main/java/br/unb/cic/analysis/AbstractMergeConflictDefinition.java:27]: In between was those that were neither source nor sink, but this approach didn't work out because I couldn't mark statements outside from the source method. I forgot to remove this attribute, will remove it. @pauloborba

File: src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java (hunk 3 above)
Comment (rafaelmotaalves) [reply src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java:11]: yes, it is, i was following the pattern I saw on the rest of the project adding the tests to the same package as the tested analysis but on the src/test folder, should it move it? @pauloborba

File: src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java (hunk 5 above)
Comment (rafaelmotaalves) [reply src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:96]: Yes, but this line is to guarantee that units outside of the method body are not considered sink, added this because if there is a flow from a source to a method call and the unit corresponding to the method definition is marked as sink, SVFA will return a flow to this method definition. @pauloborba

File: src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java (hunk 6 above)
Comment (rafaelmotaalves) [reply src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:95]: This is to check if a unit is neither sink nor source, in the case the method is source and sink, the first condition would be false and the method would return false, why would this break the logic? @pauloborba

Comment (rafaelmotaalves) [reply src/main/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysis.java:95]: what's the expected behaviour when a unit is both source and sink?

File: src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java (hunk 3 above)
Comment (rafaelmotaalves) [reply src/test/java/br/unb/cic/analysis/svfa/confluence/SVFAConfluenceAnalysisWithMultipleConfluences.java:11]: Just saw it, it was on the wrong folder
